yt_dlp/utils.py

```diff
@@ -5298,3 +5301,75 @@ def all_args(self):
 
     def parse_args(self):
         return self._parser.parse_args(list(self.all_args))
+
+
+# taken from https://github.com/python/cpython/blob/3.9/Lib/asyncio/runners.py with modifications
+def run_with_loop(main, loop):
+    if asyncio.events._get_running_loop() is not None:
+        raise RuntimeError(
+            "asyncio.run() cannot be called from a running event loop")
+
+    if not asyncio.coroutines.iscoroutine(main):
+        raise ValueError("a coroutine was expected, got {!r}".format(main))
+
+    try:
+        # asyncio.events.set_event_loop(loop)
+        return loop.run_until_complete(main)
+    finally:
+        loop.run_until_complete(loop.shutdown_asyncgens())
+        if hasattr(loop, 'shutdown_default_executor'):
+            loop.run_until_complete(loop.shutdown_default_executor())
+
+
+def _cancel_all_tasks(loop):
+    to_cancel = compat_asyncio_all_tasks(loop)
+
+    if not to_cancel:
+        return
+
+    for task in to_cancel:
+        task.cancel()
+
+    loop.run_until_complete(
+        asyncio.tasks.gather(*to_cancel, loop=loop, return_exceptions=True))
+
+    for task in to_cancel:
+        if task.cancelled():
+            continue
+        if task.exception() is not None:
+            loop.call_exception_handler({
+                'message': 'unhandled exception during asyncio.run() shutdown',
+                'exception': task.exception(),
+                'task': task,
+            })
+
+
+class WebSocketsWrapper():
+    "Wraps websockets module to use in non-async scopes"
+
+    def __init__(self, url, headers=None):
+        # self.loop = asyncio.events.get_event_loop() or asyncio.events.new_event_loop()
+        self.loop = asyncio.events.new_event_loop()
+        self.conn = compat_websockets.connect(
```

Comment:
shouldn't we throw a custom error before this if websockets is not found?

Reply:
If all calling code checks for has_websockets, this is a non-issue

Reply:
#2418 isn't checking `has_websockets` now (because it was based on the first commit of this PR), but will be fixed to check it.
I wanted to avoid making any requests before checking existence of WS library